src/app/InteractionModelEngine.cpp

```diff
@@ -764,34 +867,132 @@ CHIP_ERROR InteractionModelEngine::CanEstablishReadTransaction(const ReadHandler
+    const size_t eventPathCap     = allowUnlimited ? SIZE_MAX : GetPathPoolCapacityForReads();
+    const size_t readHandlerCap   = allowUnlimited ? SIZE_MAX : GetReadHandlerPoolCapacityForReads();
 
-    // It is safe to use & here since this function will be called on current stack.
-    if (apReadHandler->GetAttributePathCount() > kReservedPathsPerReadRequest ||
-        apReadHandler->GetEventPathCount() > kReservedPathsPerReadRequest ||
-        apReadHandler->GetDataVersionFilterCount() > kReservedPathsPerReadRequest)
+    const size_t guaranteedReadRequestsPerFabric = GetGuaranteedReadRequestsPerFabric();
+
+    size_t usedAttributePaths = 0;
+    size_t usedEventPaths     = 0;
+    size_t usedReadHandlers   = 0;
+
+    auto countResourceUsage = [&]() {
+        usedAttributePaths = 0;
+        usedEventPaths     = 0;
+        usedReadHandlers   = 0;
+        mReadHandlers.ForEachActiveObject([&](auto * handler) {
+            if (!handler->IsType(ReadHandler::InteractionType::Read))
+            {
+                return Loop::Continue;
+            }
+            usedAttributePaths += handler->GetAttributePathCount();
+            usedEventPaths += handler->GetEventPathCount();
+            usedReadHandlers++;
+            return Loop::Continue;
+        });
+    };
+
+    countResourceUsage();
+
+    if (usedAttributePaths + aRequestedAttributePathCount <= attributePathCap &&
+        usedEventPaths + aRequestedEventPathCount <= eventPathCap && usedReadHandlers < readHandlerCap)
+    {
+        // We have enough resources, then we serve the requests in a best-effort manner.
+        return Protocols::InteractionModel::Status::Success;
+    }
+
+    if ((aRequestedAttributePathCount > kMinSupportedPathsPerReadRequest &&
+         usedAttributePaths + aRequestedAttributePathCount > attributePathCap) ||
+        (aRequestedEventPathCount > kMinSupportedPathsPerReadRequest && usedEventPaths + aRequestedEventPathCount > eventPathCap))
     {
-        // Doesn't matter how many other reads are in progress; we are not
-        // going to handle this request.
-        return (allowUnlimited ? CHIP_NO_ERROR : CHIP_IM_GLOBAL_STATUS(PathsExhausted));
+        // We cannot offer enough resources, and the read transaction is requesting more than the spec limit.
+        return Protocols::InteractionModel::Status::PathsExhausted;
     }
 
-    mReadHandlers.ForEachActiveObject([&](ReadHandler * handler) {
-        if (handler->GetAccessingFabricIndex() == currentFabricIndex && handler->IsType(ReadHandler::InteractionType::Read))
+    // If we have commissioned CHIP_CONFIG_MAX_FABRICS already, and this transaction don't have an associated fabric index, reject
+    // the request if we don't have sufficient resources for this request.
+    if (mpFabricTable->FabricCount() == GetConfigMaxFabrics() && aFabricIndex == kUndefinedFabricIndex)
+    {
+        return Protocols::InteractionModel::Status::ResourceExhausted;
```

Comment:
So seems we should never return ResourcerExhausted?

It makes sense since all read transactions are expected to be finished in the near future.

Reply:
I think we do want to return ResourceExhausted for things that are using too many paths.

For PASE, it's not so much about what we return here, it's about whether PASE has any completion gaurantees on a read....  I think we will probably want to end up with such a guarantee. But again, followup for this, please.

Reply:
Per spec:
> BUSY in all other recoverable resource exhausted situations

All resources in the pool are recoverable (since they will be released at the end of transaction). We should reply Busy here.